### ISSUE TITLE
CUDA API accepts non-const string arguments when it shouldn't

### ISSUE THREAD

**Issue by max-ruttenberg** [bespoke-silicon-group/bsg_f1#282]
Any `char*` argument that is meant to be a file name, device name, or anything thing of that sort needs to be made `const`.  Actually, ALL pointer arguments should be made `const` unless you know for an absolute fact that you must modify the memory that the argument points to.

It's important that they be `const` because C/C++ allows for implicit conversion from non-const to const pointers but NOT the other way around. And for good reason; `const` is a promise to the caller that the memory will still be valid and unchanged after the function returns. 

That's a very nice promise to have and an easy one to keep in most cases... especially when arguments are meant to be string constant that you're just going to copy or forget about anyway. 

TL;DR make arguments to your library functions `const` unless you're pretty sure that they must be modified (e.g. a pointer argument that's really meant to be an output such as the `eva` argument in `hb_mc_loader_symbol_to_eva()`). But when in doubt, err on the side of const.

**Comment by bornaehsani** [comment]
This makes a lot of sense, and I really appreciate your detailed explanation. I tried to abide by it and implement it as much as I can. Thanks a lot Max!
Addressed in PR #309 .
